crates/uv-auth/src/credentials.rs

```diff
@@ -1,117 +1,156 @@
+        let host = url.host_str()?;
+        let entry = netrc
+            .hosts
+            .get(host)
+            .or_else(|| netrc.hosts.get("default"))?;
+
+        // Ensure the username matches if provided
+        if username.is_some_and(|username| username != entry.login) {
+            return None;
+        };
+
+        Some(Credentials {
+            username: Some(entry.login.clone()),
+            password: Some(entry.password.clone()),
+        })
+    }
+
+    /// Parse [`Credentials`] from a URL, if any.
+    ///
+    /// Returns [`None`] if both [`Url::username`] and [`Url::password`] are not populated.
     pub fn from_url(url: &Url) -> Option<Self> {
         if url.username().is_empty() && url.password().is_none() {
             return None;
         }
         Some(Self {
             // Remove percent-encoding from URL credentials
             // See <https://github.com/pypa/pip/blob/06d21db4ff1ab69665c22a88718a4ea9757ca293/src/pip/_internal/utils/misc.py#L497-L499>
-            username: urlencoding::decode(url.username())
-                .expect("An encoded username should always decode")
-                .into_owned(),
+            username: if url.username().is_empty() {
+                None
+            } else {
+                Some(
+                    urlencoding::decode(url.username())
+                        .expect("An encoded username should always decode")
+                        .into_owned(),
+                )
+            },
             password: url.password().map(|password| {
                 urlencoding::decode(password)
                     .expect("An encoded password should always decode")
                     .into_owned()
             }),
         })
     }
-}
 
-impl From<Authenticator> for Credentials {
-    fn from(auth: Authenticator) -> Self {
-        Credentials {
-            username: auth.login,
-            password: Some(auth.password),
+    /// Parse [`Credentials`] from an HTTP request, if any.
+    ///
+    /// Only HTTP Basic Authentication is supported.
+    pub fn from_request(request: &Request) -> Option<Self> {
+        // First, attempt to retrieve the credentials from the URL
+        Self::from_url(request.url()).or(
+            // Then, attempt to pull the credentials from the headers
+            request
+                .headers()
+                .get(reqwest::header::AUTHORIZATION)
+                .map(Self::from_header_value)?,
+        )
+    }
+
+    /// Parse [`Credentials`] from an authorization header, if any.
+    ///
+    /// Only HTTP Basic Authentication is supported.
+    ///
+    /// [`None`] will be returned if any error is encountered.
+    pub(crate) fn from_header_value(header: &HeaderValue) -> Option<Self> {
+        let mut value = header.as_bytes().strip_prefix(b"Basic ")?;
+        let mut decoder = DecoderReader::new(&mut value, &BASE64_STANDARD);
+        let mut buf = String::new();
+        decoder.read_to_string(&mut buf).ok()?;
```

Comment:
Wonder if this should `.expect`?

Reply:
Right now if we see anything we don't recognize we just give up. I'm not sure if basic authentication is _guaranteed_ to always be base64 encoded, I could look that up.